pkg/allocator/allocator.go

```diff
@@ -725,6 +754,54 @@ func (a *Allocator) Allocate(ctx context.Context, key AllocatorKey) (idpool.ID,
 	return 0, false, false, err
 }
 
+func (a *Allocator) GetWithRetry(ctx context.Context, key AllocatorKey) (idpool.ID, error) {
+	getID := func() (idpool.ID, error) {
+		id, err := a.Get(ctx, key)
+		if err != nil {
+			return 0, err
+		}
+
+		if id == idpool.NoID {
+			return 0, fmt.Errorf("security identity not found for key %s", key.String())
+		}
+
+		return id, nil
+	}
+
+	// Make a copy of the template and customize it.
+	boff := a.backoffTemplate
+	boff.Name = key.String()
+
+	var id idpool.ID
+	var err error
+
+	for attempt := 0; attempt < a.maxAllocAttempts; attempt++ {
+		id, err = getID()
+		if err == nil {
+			return id, nil
+		}
+
+		scopedLog := log.WithFields(logrus.Fields{
+			fieldKey:          key,
+			logfields.Attempt: attempt,
+		})
+
+		select {
+		case <-ctx.Done():
+			scopedLog.WithError(ctx.Err()).Warning("Ongoing key allocation has been cancelled")
+			return 0, fmt.Errorf("key allocation cancelled: %w", ctx.Err())
+		default:
+			scopedLog.WithError(err).Warning("Key allocation attempt failed")
```

Comment:
This shouldn't be a warning, as occasionally losing the race is probably to be expected. Instead, when we need to retry, it should probably be a `Debug("CiliumIdentity not yet created by operator, retrying...")` and the failure below should have more context.